### ISSUE TITLE
persistent_data_loader_workers causes dataset batch duplication / dropout.

### ISSUE THREAD

**Issue by mcmanus82** [kohya-ss/musubi-tuner#283]
When using --persistent_data_loader_workers (or --max_data_loader_n_workers X) some dataset batches get duplicated while others get dropped.
Everything was tested on a clean unmodified version of Musubi-tuner in addition to the version i'm working on for val loss to rule out any issues by my modifications.
Whit a dataset of 20-40 images this is mostly not noticeable, here is an example:

On a dataset with 40 images with repeat = 1
Latent(1)_hash (step2) and Latent(26)_hash (step27) is identical

[DEBUG] Latent(0)_hash = 2c1f3a5e9489b27b539759724be8cff3
steps:   0%|| 1/40000 [00:06<76:42:01,  6.90s/it, avr_loss=0.0904][DEBUG] Latent(1)_hash = **8327adaaa8c7e70ff9e7f7ceec47ca62**
steps:   0%|| 2/40000 [00:07<40:18:44,  3.63s/it, avr_loss=0.0924][DEBUG] Latent(2)_hash = 6f805faa8c97900efcf87c7b78c4ffb2
steps:   0%|| 3/40000 [00:07<28:15:56,  2.54s/it, avr_loss=0.297][DEBUG] Latent(3)_hash = 41a82ce1167aac36b63e500484c66c53
steps:   0%|| 4/40000 [00:07<22:09:29,  1.99s/it, avr_loss=0.278][DEBUG] Latent(4)_hash = be9abdf646fb1c1e9c3d98f1abeb91e2
steps:   0%|| 5/40000 [00:08<18:30:58,  1.67s/it, avr_loss=0.3][DEBUG] Latent(5)_hash = 6c5b8a70d2f40ef0fd6bcf08915e245f
steps:   0%|| 6/40000 [00:08<16:05:18,  1.45s/it, avr_loss=0.264][DEBUG] Latent(6)_hash = 262aaed6d0737553f625204ea0827731
steps:   0%|| 7/40000 [00:09<14:21:12,  1.29s/it, avr_loss=0.234][DEBUG] Latent(7)_hash = 3c73b38be639bef40d5aed2bf1b40a30
steps:   0%|| 8/40000 [00:09<13:02:16,  1.17s/it, avr_loss=0.242][DEBUG] Latent(8)_hash = f2dcc04a4c1ef5c852eb08be28f3fec9
steps:   0%|| 9/40000 [00:09<12:01:34,  1.08s/it, avr_loss=0.228][DEBUG] Latent(9)_hash = e15ca77abb268b138746530852633afa
steps:   0%|| 10/40000 [00:10<11:13:04,  1.01s/it, avr_loss=0.213][DEBUG] Latent(10)_hash = 6999b147af0bde96408439ff36e4c191
steps:   0%|| 11/40000 [00:10<10:33:13,  1.05it/s, avr_loss=0.211][DEBUG] Latent(11)_hash = 7f109a928732a5641e7e35a185e2687b
steps:   0%|| 12/40000 [00:10<9:59:41,  1.11it/s, avr_loss=0.198][DEBUG] Latent(12)_hash = c85af284ef14e1ae077b0e8d6b209d8a
steps:   0%|| 13/40000 [00:11<9:31:04,  1.17it/s, avr_loss=0.19][DEBUG] Latent(13)_hash = 724eda2bff1d72b4dc8d29ebcc7c6155
steps:   0%|| 14/40000 [00:11<9:07:44,  1.22it/s, avr_loss=0.184][DEBUG] Latent(14)_hash = 6cd9509793000d6cfea6793cd0d69eb9
steps:   0%|| 15/40000 [00:11<8:46:59,  1.26it/s, avr_loss=0.176][DEBUG] Latent(15)_hash = 8ece559c37719a526429df3c7284e871
steps:   0%|| 16/40000 [00:12<8:28:41,  1.31it/s, avr_loss=0.172][DEBUG] Latent(16)_hash = b3d6882df394b5fa0ee7b30a796da8a8
steps:   0%|| 17/40000 [00:12<8:12:45,  1.35it/s, avr_loss=0.167][DEBUG] Latent(17)_hash = c0fef8a3ce1b6d1a174ab22d99646da3
steps:   0%|| 18/40000 [00:12<7:58:05,  1.39it/s, avr_loss=0.162][DEBUG] Latent(18)_hash = e1e76882a95a942d8483bf660b3016a0
steps:   0%|| 19/40000 [00:13<7:45:38,  1.43it/s, avr_loss=0.158][DEBUG] Latent(19)_hash = df46242a04dd386688559793cfe9a376
steps:   0%|| 20/40000 [00:13<7:33:49,  1.47it/s, avr_loss=0.153][DEBUG] Latent(20)_hash = f601115da4692b3e20b0e7660a269c9f
steps:   0%|| 21/40000 [00:13<7:23:27,  1.50it/s, avr_loss=0.153][DEBUG] Latent(21)_hash = 2c1f3a5e9489b27b539759724be8cff3
steps:   0%|| 22/40000 [00:14<7:13:51,  1.54it/s, avr_loss=0.15][DEBUG] Latent(22)_hash = 698e2af7d104ede5d7d8c7569fb32a21
steps:   0%|| 23/40000 [00:14<7:05:16,  1.57it/s, avr_loss=0.146][DEBUG] Latent(23)_hash = 3f5308c9e9d6efe24bd90dc167578a45
steps:   0%|| 24/40000 [00:15<6:57:01,  1.60it/s, avr_loss=0.143][DEBUG] Latent(24)_hash = 61c299a1d8860bc089b2e673ebd21d7f
steps:   0%|| 25/40000 [00:15<6:49:59,  1.63it/s, avr_loss=0.147][DEBUG] Latent(25)_hash = 27507b8f5aa4ed7f31260e8f742e8f52
steps:   0%|| 26/40000 [00:15<6:42:58,  1.65it/s, avr_loss=0.143][DEBUG] Latent(26)_hash = **8327adaaa8c7e70ff9e7f7ceec47ca62**
....

With smaller datasets, lets say a validation dataset with 5 images this becomes extremely problematic, again example:

The first epoch has two pars of duplicates to begin with, the second epoch "only" has one, the third has two pairs again. Not to mention that shuffle=flase is not respected at all, and thus deterministic validation becomes impossible to achieve.

[DEBUG] Latent(0)_hash = **41a82ce1167aac36b63e500484c66c53**
steps:   0%|| 1/5000 [00:04<5:47:18,  4.17s/it, avr_loss=0.0542][DEBUG] Latent(1)_hash = _61c299a1d8860bc089b2e673ebd21d7f_
steps:   0%|| 2/5000 [00:04<3:08:58,  2.27s/it, avr_loss=0.0511][DEBUG] Latent(2)_hash = _61c299a1d8860bc089b2e673ebd21d7f_
steps:   0%|| 3/5000 [00:04<2:15:45,  1.63s/it, avr_loss=0.276][DEBUG] Latent(3)_hash = **41a82ce1167aac36b63e500484c66c53**
steps:   0%|| 4/5000 [00:05<1:49:09,  1.31s/it, avr_loss=0.262][DEBUG] Latent(4)_hash = 698e2af7d104ede5d7d8c7569fb32a21
steps:   0%|| 5/5000 [00:05<1:33:11,  1.12s/it, avr_loss=0.271]

[DEBUG] Latent(0)_hash = **61c299a1d8860bc089b2e673ebd21d7f**
steps:   0%|| 6/5000 [00:06<1:32:45,  1.11s/it, avr_loss=0.272][DEBUG] Latent(1)_hash = 698e2af7d104ede5d7d8c7569fb32a21
steps:   0%|| 7/5000 [00:07<1:23:48,  1.01s/it, avr_loss=0.273][DEBUG] Latent(2)_hash = 262aaed6d0737553f625204ea0827731
steps:   0%|| 8/5000 [00:07<1:16:58,  1.08it/s, avr_loss=0.191][DEBUG] Latent(3)_hash = **61c299a1d8860bc089b2e673ebd21d7f**
steps:   0%|| 9/5000 [00:07<1:11:40,  1.16it/s, avr_loss=0.168][DEBUG] Latent(4)_hash = 41a82ce1167aac36b63e500484c66c53
steps:   0%|| 10/5000 [00:08<1:07:25,  1.23it/s, avr_loss=0.131]

[DEBUG] Latent(0)_hash = **61c299a1d8860bc089b2e673ebd21d7f**
steps:   0%|| 11/5000 [00:09<1:08:50,  1.21it/s, avr_loss=0.163][DEBUG] Latent(1)_hash = _262aaed6d0737553f625204ea0827731_
steps:   0%|| 12/5000 [00:09<1:05:53,  1.26it/s, avr_loss=0.165][DEBUG] Latent(2)_hash = **61c299a1d8860bc089b2e673ebd21d7f**
steps:   0%|| 13/5000 [00:09<1:03:03,  1.32it/s, avr_loss=0.113][DEBUG] Latent(3)_hash = 41a82ce1167aac36b63e500484c66c53
steps:   0%|| 14/5000 [00:10<1:00:38,  1.37it/s, avr_loss=0.108][DEBUG] Latent(4)_hash = _262aaed6d0737553f625204ea0827731_
steps:   0%|| 15/5000 [00:10<58:32,  1.42it/s, avr_loss=0.0934]

Not enabling persistent_data_loader_workers solves duplication / dropout issue.
The main drawback is that if not enabled, worker count defaults to max cpu cores, which significantly increases training time. Defaulting to 2 would be preferable.

**Comment by Sarania** [comment]
Wow nice find! I've had some issues in this area recently as well. When I try to encode my datasets with the default number of workers nproc-1 (which is 15 on my system) it will lock up, ultimately spinning right here:

```        def aggregate_future(consume_all: bool = False):
            while len(futures) >= num_workers or (consume_all and len(futures) > 0):
                completed_futures = [future for future in futures if future.done()]
                if len(completed_futures) == 0:
                    if len(futures) >= num_workers or consume_all:  # to avoid adding too many futures
                        time.sleep(0.1)  # <--- here at dataset/image_video_dataset.py line 1577
                        continue
                    else:
                        break  # submit batch if possible
```

This only started happening recently and I wonder if it's related to the issue you are seeing? If I set num_workers = 2 or 4 it completes fine. I know it's not exactly the same but it seemed worth mentioning. I hadn't brought it up before because I wasn't sure if it was some quirk of my own changes, though I haven't interacted with image_video_dataset.py really.

**Comment by mcmanus82** [comment]
Thanks :) I honestly don't know if these are related. For me the issue occurs even if i only pass the --max_data_loader_n_workers X arg with any number. I think that arg alone also enables persistent workers, at least that's what i figured from a quick glance at the code.
I generally had a bad experience with persistent workers, in SD-Scripts it gave me random crashes and errors every 10-30 minutes.

If you have a couple of minutes, could i ask you to check the discussion i made about reviving val loss? I'm pretty much out of ideas why it gives "useless" data curves.

**Comment by Sarania** [comment]
> This issue has been fixed, sorry for the inconvenience. If the issue persists please reopen this.
> 
> This seems to be a different problem from the one Sarania mentioned.
> 
> The default number of workers is `os.cpu_count() + 4`, so `os.cpu_count() - 1` is a conservative number, but if problems occur it may be better to reduce it.

The more I thought about it the more I realized it was likely different yeah, sorry about that. I wouldn't worry about mine too much unless someone else brings it up, it's easily fixed by setting num_workers to 4 and it's not impossible that I've caused it myself XD
